x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -107,61 +103,49 @@ function getSpanItem(span: Span): PartialSpanItem {
   };
 }
 
-export function getWaterfallRoot(
-  items: Array<PartialSpanItem | IWaterfallItemTransaction>,
+export function getWaterfallItems(
+  items: IWaterfallItem[],
   entryTransactionItem: IWaterfallItem
 ) {
-  const itemsByParentId = groupBy(
+  const itemsById: IWaterfallIndex = indexBy(items, 'id');
+  const childrenByParentId = groupBy(
     items,
     item => (item.parentId ? item.parentId : 'root')
   );
-  const itemsById: IWaterfallIndex = {};
-
-  const itemsByTransactionId = indexBy(
-    items.filter(item => item.docType === 'transaction'),
-    item => item.id
-  ) as { [key: string]: IWaterfallItemTransaction };
-
-  function getWithChildren(
-    item: PartialSpanItem | IWaterfallItemTransaction
-  ): IWaterfallItem {
-    const children = itemsByParentId[item.id] || [];
-    const nextChildren = sortBy(children, 'timestamp').map(getWithChildren);
-    let fullItem;
-
-    // add parent transaction to spans
-    if (item.docType === 'span') {
-      fullItem = {
-        parentTransaction:
-          itemsByTransactionId[item.span.transaction.id].transaction,
-        ...item,
-        offset: item.timestamp - entryTransactionItem.timestamp,
-        children: nextChildren
-      };
-    } else {
-      fullItem = {
-        ...item,
-        offset: item.timestamp - entryTransactionItem.timestamp,
-        children: nextChildren
-      };
-    }
 
-    // TODO: Think about storing this tree as a single, flat, indexed structure
-    // with "children" being an array of ids, instead of it being a real tree
-    itemsById[item.id] = fullItem;
+  function getOrderedItems(item: IWaterfallItem): IWaterfallItem[] {
+    const children = sortBy(childrenByParentId[item.id] || [], 'timestamp');
 
-    return fullItem;
+    // mutating to ensure both data structures ("itemsById" and "items") have the same data
+    item.childIds = children.map(child => child.id);
+    item.offset = item.timestamp - entryTransactionItem.timestamp;
+
+    const deepChildren = flatten(children.map(getOrderedItems));
```

Comment:
Sucks that Array#flatMap isn't supported in IE or edge, but you could maybe use the lodash `flatMap` for now* if you want to slim this down to one method:
```js
const deepChildren = _.flatMap(children, getOrderedItems)
```

*if lodash `flatMap` is in our version of lodash, not sure :/

Reply:
Oh, that's nice. Will look into that. Do you think it's worth pulling in `lodash.flatMap` if we only have one usage?

Reply:
Meh, probably not. Old browsers and old lodash!